### PR TITLE
Enumerable re-implementation

### PR DESCRIPTION
**Pull request by aldemeery**
### Added
- Re-implementation for:
    - `#each` method.
    - `#each_with_index` method.
    - `#select` method.
    - `#all?` method.
    - `#any?` method.
    - `#none?` method.
    - `#count` method.
    - `#my_map` method.
    - `#my_map!` method.
    - `#my_inject` method.

### Fixed
- Rubocop errors.

### CODE REVIEW

File: main.rb [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+                    end
+      end
+    elsif block_given?
+      my_each { |item| all_true &= yield(item) }
+    else
+      all_true = my_all? { |item| item }
+    end
+
+    all_true
+  end
```
Comment (Dandush03) [inline main.rb:42]: ### #my_all?
When testing this method, each time that I give it an argument I received the following` TypeError: Class or module required.`
here I'll add an example of one of the tests:
``` ruby
irb> arr = Array.new(10) { rand(0...5) }
irb> arr.my_all?(3)
  => TypeError (class or module required)
```
> expected to be false

File: main.rb [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+      puts 'Warning, the block is being ignored. ' if block
+
+      my_each { |item| result |= (arg.is_a?(Regexp) ? arg === item.to_s : item.is_a?(arg)); }
+    elsif block_given?
+      my_each { |item| result |= yield(item) }
+    else
+      return my_any? { |item| item }
+    end
+    result
+  end
```
Comment (Dandush03) [inline main.rb:56]: ### #my_any?
When testing this method, each time that I give it an argument I received the following` TypeError: Class or module required.`
here I'll add an example of one of the tests:
``` ruby
irb> arr = %w[dog door rod blade] 
irb> arr.my_any?('cat')
  => TypeError (class or module required)
```
> expected to be false

File: main.rb [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+  def my_none?(*args, &block)
+    !my_any?(*args, &block)
+  end
```
Comment (Dandush03) [inline main.rb:60]: ### #my_none?
When testing this method, each time that I give it an argument I received the following` TypeError: Class or module required.`
here I'll add an example of one of the tests:
``` ruby
irb> arr = %w[dog door rod blade] 
irb> arr.my_none?(5)
  => TypeError (class or module required)
```
> expected to be false

File: main.rb [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+  def my_count?(*args, &block)
```
Comment (Dandush03) [inline main.rb:62]: ### #my_count
Question marks are used when expecting a boolean value

File: main.rb [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    counter
+  end
+
+  def my_map(*args)
+    return enum_for(__method__) unless block_given? || args[0].nil?
+
+    new_array = []
+    my_each { |item| new_array << args[0].nil? ? yield(item) : args[0].call(item) }
+    new_array
+  end
```
Comment (Dandush03) [inline main.rb:82]: ### #my_map
Should return an enumerator if no block is given

File: main.rb [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    elsif args.size == 1 && !block_given?
+      memo = first
+      my_each { |item| memo = memo.send(args[0], item) }
+    else
+      memo = args[0] || first
+      my_each { |item| memo = yield(memo, item) if block_given? }
+    end
+    memo
+  end
+end
```
Comment (Dandush03) [inline main.rb:104]: ### #my_inject?
enumerable my_inject when a symbol is specified combines each element of the collection by applying the symbol as a named method
``` ruby
irb> arr = Array.new(10) { rand(0...5) }
irb> arr.my_inject(:+) == arr.inject(:+)
  => false
```
> expected to be true

**Review by Dandush03 — COMMENTED** [review]
# Status: 🟠 Required Changes 🟠 
## Hey! 👋 
### Great Work so far, but we still need some changes to meets all the requirements
#### Please check out my comments below and then submit for a new review
## Happy Coding 💻 
Daniel Laloush

File: main.rb (hunk 1 above)
Comment (aldemeery) [reply main.rb]: Fixed

File: main.rb (hunk 2 above)
Comment (simandebvu) [reply main.rb]: Fixed

File: main.rb (hunk 4 above)
Comment (aldemeery) [reply main.rb]: Fixed

File: main.rb (hunk 3 above)
Comment (aldemeery) [reply main.rb:60]: Fixed

File: main.rb (hunk 6 above)
Comment (aldemeery) [reply main.rb]: Fixed

File: main.rb [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+  end
+
+  def my_map!(*args)
+    return enum_for(__method__) if !block_given? && args[0].nil?
+
+    my_each_with_index { |item, index| self[index] = (args[0].nil? ? yield(item) : args[0].call(item)) }
+    self
+  end
+
+  def my_inject(*args)
```
Comment (serdg0) [inline main.rb:91]: My inject should:

- [ ] Properly combine the elements when passed a block with a binary operation. Here I compared your method with the original one, with the block { proc { |sum, n| sum + n } } passed like array.my_inject(&block) and got a different result.

- [ ] It should **not** mutate the original array.

- [ ] The same as the 1st point, but passing a symbol. Here I compared with array.my_inject(:+) and got a different result from the original one.

**Review by serdg0 — CHANGES_REQUESTED** [review]
Hello @aldemeery @simandebvu 

Great job! You guys are very close to finish! :racing_car: :racing_car: 

All of your methods work as expected, but 1. And by the work you've done, it won't be hard for you to make this project outstanding.

When done fixing the issues, submit for another CRR.

**Happy Coding!** :computer: :nerd_face:
**_Sergio Diaz_**

File: main.rb (hunk 7 above)
Comment (aldemeery) [reply main.rb:91]: Fixed

File: main.rb [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+      first_el = shift
+      memo = args[0] || first_el
+      my_each { |item| memo = yield(memo, item) if block_given? }
+      unshift(first_el)
+
+      return memo
+    end
+    memo
+  end
+end
```
Comment (euqueme) [inline main.rb:112]: - [ ] should have working `my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

- my_inject should raise a "LocalJumpError" when no block or argument is given
```ruby
[1,5,10,20].my_inject == [1,5,10,20].inject
```
Yours (incorrect)
image

Real inject (correct)
image

- my_inject, when a block is given without an initial value, combines all elements of enum by applying a binary operation, specified by a block::Range
```ruby
(1..20).my_inject { |prod, n| prod * n } == (1..20).inject { |prod, n| prod * n }
```
Yours (incorrect)
image

Real inject (correct)
image

- my_inject, when a block is given with an initial value, combines all elements of enum by applying a binary operation, specified by a block
```ruby
(1..20).my_inject(4) { |prod, n| prod * n } == (1..20).inject(4) { |prod, n| prod * n }
```
Yours (incorrect)
image

Real inject (correct)
image

- my_inject, when a symbol is specified without an initial value, combines each element of the collection by applying the symbol as a named method:Range
```ruby
(1..20).my_inject(:*) == (1..20).inject(:*)
```
Yours (incorrect)
image

Real inject (correct)
image

**Review by euqueme — CHANGES_REQUESTED** [review]
> ### Pull Request under the review by Maria Eugenia Queme P. (TSE)

### Hello @aldemeery and @simandebvu  :lollipop: :balloon: 

### Thank you for working on the changes requested by the previous reviewer :medal_sports: :medal_sports: your enumerable methods are working well and you should feel very proud of the work you've done, there are still a couple of changes yet to make to your my_inject so please look at the comments I've left in your code below :point_down: 

### That's it, team! :raised_hands: 

### Please :pray: do the changes :technologist: :technologist: and submit back for a code review when you are done :memo: 

Happy coding! :rocket: 

Any question? ask me on slack ** @ Maru** :woman_technologist: 

PROJECT STATUS: REQUESTED CHANGES :vertical_traffic_light:

File: main.rb (hunk 8 above)
Comment (aldemeery) [reply main.rb:112]: Fixed

**Review by MarioDena — APPROVED** [review]
Hello great work on your project!! :medal_military: everything is working correctly you can go ahead and merge this PR

## Project approved

Cheers! :smile:

**Pull request merged**